src/Catalyst.Core.Modules.Dfs/DfsService.cs

```diff
@@ -0,0 +1,529 @@
+                    await randomWalk.StartAsync().ConfigureAwait(false);
+                }
+            };
+            Log.Debug("waiting for discovery services to start");
+            await Task.WhenAll(tasks.Select(t => t())).ConfigureAwait(false);
+
+            multicast?.Start();
+
+            Log.Debug("started");
+        }
+
+        /// <summary>
+        ///     Stops the running services.
+        /// </summary>
+        /// <returns>
+        ///     A task that represents the asynchronous operation.
+        /// </returns>
+        /// <remarks>
+        ///     Multiple calls are okay.
+        /// </remarks>
+        public async Task StopAsync()
+        {
+            Log.Debug("stopping");
+            try
+            {
+                var tasks = _stopTasks.ToArray();
+                _stopTasks = new ConcurrentBag<Func<Task>>();
+                await Task.WhenAll(tasks.Select(t => t())).ConfigureAwait(false);
+            }
+            catch (Exception e)
+            {
+                Log.Error("Failure when stopping the engine", e);
+            }
+
+            // Many services use cancellation to stop.  A cancellation may not run
+            // immediately, so we need to give them some.
+            // TODO: Would be nice to make this deterministic.
+            await Task.Delay(TimeSpan.FromMilliseconds(100)).ConfigureAwait(false);
+
+            Log.Debug("stopped");
+        }
+
+        /// <summary>
+        ///     A synchronous start.
+        /// </summary>
+        /// <remarks>
+        ///     Calls <see cref="StartAsync" /> and waits for it to complete.
+        /// </remarks>
+        public void Start() { StartAsync().ConfigureAwait(false).GetAwaiter().GetResult(); }
+
+        /// <summary>
+        ///     A synchronous stop.
+        /// </summary>
+        /// <remarks>
+        ///     Calls <see cref="StopAsync" /> and waits for it to complete.
+        /// </remarks>
+        public void Stop()
+        {
+            Log.Debug("stopping");
+            try
+            {
+                var tasks = _stopTasks.ToArray();
+                _stopTasks = new ConcurrentBag<Func<Task>>();
+                foreach (var task in tasks)
+                {
+                    task().ConfigureAwait(false).GetAwaiter().GetResult();
+                }
+            }
+            catch (Exception e)
+            {
+                Log.Error("Failure when stopping the engine", e);
+            }
+        }
+
+#pragma warning disable VSTHRD100 // Avoid async void methods
+        /// <summary>
+        ///     Fired when a peer is discovered.
+        /// </summary>
+        /// <param name="sender"></param>
+        /// <param name="peer"></param>
+        /// <remarks>
+        ///     Registers the peer with the <see cref="SwarmService" />.
+        /// </remarks>
+        private void OnPeerDiscovered(object sender, Peer peer)
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Remove this unused method parameter 'sender'.](https://app.codacy.com/manual/catalyst-network/Catalyst.Framework/pullRequest?prid=5051852)